test/blackbox/common/BlackboxTestsDiscovery.cpp

```diff
@@ -2131,3 +2141,69 @@ TEST(Discovery, MultipleXMLProfileLoad)
     thr_reader.join();
     thr_writer.join();
 }
+
+//! Regression test for redmine issue 20641
+TEST(Discovery, discovery_cyclone_participant_with_custom_pid)
+{
+    using namespace eprosima::fastdds::dds;
+    using namespace eprosima::fastrtps::rtps;
+
+    /* Custom participant listener to count number of discovered participants */
+    class DiscoveryListener : public DomainParticipantListener
+    {
+    public:
+
+        void on_participant_discovery(
+                DomainParticipant*,
+                ParticipantDiscoveryInfo&& info) override
+        {
+            if (ParticipantDiscoveryInfo::DISCOVERED_PARTICIPANT == info.status)
+            {
+                discovered_participants_++;
+            }
+            else if (ParticipantDiscoveryInfo::REMOVED_PARTICIPANT == info.status)
+            {
+                discovered_participants_--;
+            }
+        }
+
+        uint8_t discovered_participants() const
+        {
+            return discovered_participants_;
+        }
+
+    private:
+
+        std::atomic<uint8_t> discovered_participants_{0};
+    };
+
+    /* Create a datagram injection transport */
+    using eprosima::fastdds::rtps::DatagramInjectionTransportDescriptor;
+    using eprosima::fastdds::rtps::DatagramInjectionTransport;
+    auto low_level_transport = std::make_shared<UDPv4TransportDescriptor>();
+    auto transport = std::make_shared<DatagramInjectionTransportDescriptor>(low_level_transport);
+
+    /* Disable builtin transport and add custom one */
+    DomainParticipantQos participant_qos = PARTICIPANT_QOS_DEFAULT;
+    participant_qos.transport().use_builtin_transports = false;
+    participant_qos.transport().user_transports.clear();
+    participant_qos.transport().user_transports.push_back(transport);
+
+    /* Create participant with custom transport and listener */
+    DiscoveryListener listener;
+    uint32_t domain_id = static_cast<uint32_t>(GET_PID()) % 230;
```

Comment:
Shouldn't be the same Domain Id as the one in the datagram?

Reply:
Yes and no, Right now, Fast DDS does not use the incoming `PID_DOMAIN_ID` for participant discovery. IMO the way to go here would be to craft a Data(p) with only the minimum set of PIDs (domain ID is not included in that set) and the offending ones. That way we'd verify exactly what we want. I decided to randomize the domain here to avoid discovering participants from elsewhere.

For 3.0, we will honor the `PID_DOMAIND_ID`, and since this is a hotfix, I've decided to do what you suggest in a follow up then.